File: mobile/app/src/main/java/com/flightrisk/app/drone/DroneManager.kt
package com.flightrisk.app.drone

import android.content.Context
import android.util.Log
import com.flightrisk.app.config.FlightRiskConfig
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Facade that owns [TelloConnection], [TelloFrameSource], and [TelloWifiChecker]
 * and coordinates them into a single lifecycle-aware entry point.
 *
 * DroneManager is created by [MainActivity] when the user taps "Connect to Drone"
 * and destroyed on disconnect or activity teardown. It monitors telemetry for
 * critical conditions (low battery, crash, connection loss, frozen stream) and
 * emits [DroneAlert]s that the UI layer can observe.
 *
 * @param context Android context for WiFi checking.
 * @param config  App-wide configuration (provides [DroneConfig]).
 */
class DroneManager(
    private val context: Context,
    private val config: FlightRiskConfig,
) {

    companion object {
        private const val TAG = "DroneManager"
        private const val MAX_AVOIDANCE_RETRIES = 5
        private const val AVOIDANCE_SIDE_CM = 30
        private const val AVOIDANCE_REVERSE_CM = 50
    }

    val wifiChecker = TelloWifiChecker(context)
    val connection = TelloConnection(config.drone)
    val frameSource = TelloFrameSource(config.drone)

    /** Observable drone state for UI binding. */
    val droneState: StateFlow<TelloState> = connection.state

    // ------------------------------------------------------------------
    // Alert system
    // ------------------------------------------------------------------

    sealed class DroneAlert {
        data object ConnectionLost : DroneAlert()
        data class BatteryCritical(val percent: Int) : DroneAlert()
        data object CrashDetected : DroneAlert()
        data object StreamFrozen : DroneAlert()
        data class ObstacleDetected(val action: String, val centerDepth: Float) : DroneAlert()
        data object MatchPause : DroneAlert()
    }

    private val _alerts = MutableSharedFlow<DroneAlert>(extraBufferCapacity = 8)
    val alerts: SharedFlow<DroneAlert> = _alerts.asSharedFlow()

    // ------------------------------------------------------------------
    // Coroutine management
    // ------------------------------------------------------------------

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var monitorJob: Job? = null
    @Volatile
    private var commandedLanding = false

    // ------------------------------------------------------------------
    // Obstacle avoidance
    // ------------------------------------------------------------------

    private var obstacleGuard: ObstacleGuard? = null

    fun initObstacleGuard() {
        if (obstacleGuard != null) return
        try {
            obstacleGuard = ObstacleGuard(context)
            if (obstacleGuard?.isAvailable == true) {
                Log.i(TAG, "Obstacle avoidance enabled")
            } else {
                Log.w(TAG, "Obstacle avoidance: MiDaS model not available")
                obstacleGuard = null
            }
        } catch (e: Exception) {
            Log.w(TAG, "Obstacle avoidance init failed: ${e.message}")
            obstacleGuard = null
        }
    }

    // ------------------------------------------------------------------
    // Connect / disconnect
    // ------------------------------------------------------------------

    suspend fun connectAndStream(): Boolean {
        val wifiStatus = wifiChecker.check()
        if (wifiStatus !is TelloWifiChecker.WifiStatus.OnTelloWifi) {
            val guidance = wifiChecker.getGuidanceMessage(wifiStatus)
            Log.w(TAG, "Not on Tello WiFi: $guidance")
            return false
        }

        val connected = connection.connect()
        if (!connected) {
            Log.e(TAG, "Connection failed")
            return false
        }

        val streaming = connection.startStream()
        if (!streaming) {
            Log.e(TAG, "Stream start failed")
            connection.disconnect()
            return false
        }

        frameSource.onStreamFrozen = { recoverStream() }
        frameSource.start()
        startMonitoring()

        // Initialize obstacle avoidance in background
        scope.launch { initObstacleGuard() }

        Log.i(TAG, "Connected and streaming")
        return true
    }

    suspend fun disconnect() {
        monitorJob?.cancel()
        monitorJob = null

        val state = droneState.value
        if (state.telemetry.isFlying) {
            Log.w(TAG, "Disconnecting while flying -- attempting to land first")
            try {
                withTimeoutOrNull(3000) {
                    connection.land()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Land attempt during disconnect failed: ${e.message}")
            }
        }

        frameSource.onStreamFrozen = null
        frameSource.stop()
        try {
            connection.stopStream()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "Error stopping stream: ${e.message}")
        }
        connection.disconnect()
        obstacleGuard?.close()
        obstacleGuard = null
        Log.i(TAG, "Disconnected")
    }

    fun shutdown() {
        scope.cancel()
        obstacleGuard?.close()
        obstacleGuard = null
        Log.i(TAG, "Shutdown complete")
    }

    // ------------------------------------------------------------------
    // Stream recovery
    // ------------------------------------------------------------------

    private fun recoverStream() {
        scope.launch {
            withContext(Dispatchers.IO) {
                Log.w(TAG, "Recovering frozen stream")
                frameSource.stop()
                connection.sendCommand("streamoff")
                delay(500)
                connection.sendCommand("streamon")
                delay(200)
                frameSource.onStreamFrozen = { recoverStream() }
                frameSource.start()
                _alerts.tryEmit(DroneAlert.StreamFrozen)
            }
        }
    }

    // ------------------------------------------------------------------
    // Lifecycle hooks
    // ------------------------------------------------------------------

    fun onActivityPause() {
        val state = droneState.value
        if (state.telemetry.isFlying) {
            Log.i(TAG, "Activity pausing while flying -- hovering")
            scope.launch { connection.hover() }
        }
    }

    fun onActivityDestroy() {
        monitorJob?.cancel()
        monitorJob = null
        val state = droneState.value
        Thread {
            frameSource.stop()
            if (state.telemetry.isFlying) {
                Log.w(TAG, "Activity destroying while flying -- emergency landing")
                try {
                    runBlocking(Dispatchers.IO) {
                        withTimeoutOrNull(1500) {
                            connection.land()
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Emergency land failed: ${e.message}")
                }
            }
            connection.disconnect()
            obstacleGuard?.close()
            scope.cancel()
        }.start()
        Log.i(TAG, "Destroyed")
    }

    // ------------------------------------------------------------------
    // Flight command delegates
    // ------------------------------------------------------------------

    suspend fun takeoff() = connection.takeoff()

    suspend fun land(): Boolean {
        commandedLanding = true
        return connection.land()
    }

    suspend fun move(direction: String, distanceCm: Int) =
        connection.move(direction, distanceCm)

    suspend fun rotate(degrees: Int) = connection.rotate(degrees)

    suspend fun rcControl(lr: Int, fb: Int, ud: Int, yaw: Int) =
        connection.rcControl(lr, fb, ud, yaw)

    suspend fun hover() = connection.hover()

    suspend fun emergencyStop(): Boolean {
        commandedLanding = true
        return connection.emergencyStop()
    }

    // ------------------------------------------------------------------
    // Pause / resume for match detection
    // ------------------------------------------------------------------

    @Volatile
    var searchPaused: Boolean = false
        private set

    /**
     * Pause the active search pattern and hover in place.
     * Called when the vision pipeline detects a confident match.
     */
    fun pauseSearchForMatch() {
        if (!searchActive || searchPaused) return
        searchPaused = true
        Log.i(TAG, "Search paused — match detected, hovering")
        scope.launch {
            try {
                connection.hover()
            } catch (e: Exception) {
                Log.w(TAG, "Hover on match pause failed: ${e.message}")
            }
        }
        _alerts.tryEmit(DroneAlert.MatchPause)
    }

    /**
     * Resume the search pattern after the operator dismisses a match alert.
     */
    fun resumeSearch() {
        if (!searchPaused) return
        searchPaused = false
        Log.i(TAG, "Search resumed")
    }

    // ------------------------------------------------------------------
    // Search pattern execution (with obstacle avoidance)
    // ------------------------------------------------------------------

    private var searchJob: Job? = null

    @Volatile
    var searchActive: Boolean = false
        private set

    @Volatile
    var searchProgress: Pair<Int, Int> = Pair(0, 0)
        private set

    fun startSearchPattern(pattern: PatternType = PatternType.EXPANDING_SQUARE) {
        if (searchActive) return
        val waypoints = SearchPattern.generate(pattern)
        searchActive = true
        searchPaused = false
        searchProgress = Pair(0, waypoints.size)
        Log.i(TAG, "Starting search pattern: ${pattern.displayName}, ${waypoints.size} waypoints")

        searchJob = scope.launch {
            for ((index, wp) in waypoints.withIndex()) {
                if (!searchActive) break

                // Wait while paused for match inspection
                while (searchPaused && searchActive) {
                    delay(200)
                }
                if (!searchActive) break

                searchProgress = Pair(index + 1, waypoints.size)

                try {
                    if (wp.distanceCm > 0) {
                        // Obstacle check before each forward move
                        if (!checkAndAvoidObstacles(index)) {
                            Log.w(TAG, "Skipping waypoint ${index + 1} — obstacle avoidance exhausted")
                            continue
                        }

                        connection.move(wp.direction, wp.distanceCm)
                        delay(500)
                    }
                    if (wp.rotateDegrees != 0) {
                        connection.rotate(wp.rotateDegrees)
                        delay(500)
                    }
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    Log.e(TAG, "Search pattern error at waypoint ${index + 1}: ${e.message}")
                    break
                }
            }

            searchActive = false
            searchPaused = false
            searchProgress = Pair(0, 0)
            Log.i(TAG, "Search pattern complete")
        }
    }

    /**
     * Check for obstacles before moving. If an obstacle is detected,
     * perform evasive maneuvers up to [MAX_AVOIDANCE_RETRIES] times.
     *
     * @return true if the path is now clear, false if retries exhausted.
     */
    private suspend fun checkAndAvoidObstacles(waypointIndex: Int): Boolean {
        val guard = obstacleGuard ?: return true

        for (retry in 1..MAX_AVOIDANCE_RETRIES) {
            if (!searchActive) return false

            val frame = frameSource.getLatestFrame() ?: return true

            val check = guard.checkPath(frame)
            if (check.safe) return true

            Log.w(TAG, "Obstacle at waypoint ${waypointIndex + 1}: " +
                "action=${check.action}, depth=${check.centerDepth}, retry=$retry")
            _alerts.tryEmit(DroneAlert.ObstacleDetected(check.action, check.centerDepth))

            if (retry >= MAX_AVOIDANCE_RETRIES) {
                Log.w(TAG, "Max avoidance retries at waypoint ${waypointIndex + 1}")
                return false
            }

            try {
                when (check.action) {
                    "go_left" -> {
                        connection.move("left", AVOIDANCE_SIDE_CM)
                        delay(500)
                    }
                    "go_right" -> {
                        connection.move("right", AVOIDANCE_SIDE_CM)
                        delay(500)
                    }
                    "reverse" -> {
                        connection.move("back", AVOIDANCE_REVERSE_CM)
                        delay(500)
                        connection.rotate(90)
                        delay(500)
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                Log.e(TAG, "Avoidance maneuver failed: ${e.message}")
                return false
            }
        }
        return false
    }

    fun stopSearchPattern() {
        searchActive = false
        searchPaused = false
        searchJob?.cancel()
        searchJob = null
        searchProgress = Pair(0, 0)
        scope.launch {
            try { connection.hover() } catch (_: Exception) {}
        }
        Log.i(TAG, "Search pattern stopped")
    }

    // ------------------------------------------------------------------
    // Telemetry monitoring
    // ------------------------------------------------------------------

    private fun startMonitoring() {
        monitorJob?.cancel()
        monitorJob = scope.launch {
            var wasConnected = true
            var wasFlying = false
            var batteryAlertSent = false

            connection.state.collect { state ->
                val isConnected = state.connectionState != TelloConnectionState.DISCONNECTED &&
                    state.connectionState != TelloConnectionState.ERROR
                if (wasConnected && !isConnected) {
                    Log.e(TAG, "Connection lost")
                    _alerts.tryEmit(DroneAlert.ConnectionLost)
                }
                wasConnected = isConnected

                val battery = state.telemetry.battery
                if (battery != null && battery in 0..config.drone.batteryCriticalThreshold && !batteryAlertSent) {
                    Log.w(TAG, "Battery critical: $battery%")
                    _alerts.tryEmit(DroneAlert.BatteryCritical(battery))
                    batteryAlertSent = true
                    if (state.telemetry.isFlying) {
                        Log.w(TAG, "Auto-landing due to critical battery")
                        land()
                    }
                }
                if (battery != null && battery > config.drone.batteryCriticalThreshold) {
                    batteryAlertSent = false
                }

                if (wasFlying && !state.telemetry.isFlying) {
                    if (state.telemetry.height == 0 && isConnected && !commandedLanding) {
                        Log.e(TAG, "Crash detected")
                        _alerts.tryEmit(DroneAlert.CrashDetected)
                    }
                    commandedLanding = false
                }
                wasFlying = state.telemetry.isFlying
            }
        }
    }
}

File: ios/FlightRisk/Drone/DroneManager.swift
import Foundation
import os.log

// MARK: - DroneAlert

/// Observable alerts emitted by DroneManager for UI consumption.
enum DroneAlert: Equatable {
    case connectionLost
    case batteryWarning(Int)
    case batteryCritical(Int)
    case crashDetected
    case streamFrozen
    case obstacleDetected(action: String, confidence: Float)
    case matchPause
}

// MARK: - DroneManager

/// Facade that owns `TelloConnection`, `TelloFrameSource`, `TelloWifiChecker`,
/// `SearchPattern`, and `ObstacleGuard` and coordinates them into a single
/// lifecycle-aware entry point.
///
/// Port of Android `DroneManager.kt` to Swift actor. Monitors telemetry for
/// critical conditions (low battery, crash, connection loss, frozen stream)
/// and emits `DroneAlert`s through an `AsyncStream`.
actor DroneManager {

    // MARK: - Constants

    private static let logger = Logger(subsystem: "com.flightrisk", category: "DroneManager")
    private static let maxAvoidanceRetries = 5
    private static let avoidanceSideCm = 30
    private static let avoidanceReverseCm = 50

    // MARK: - Sub-components

    let wifiChecker: TelloWifiChecker
    let connection: TelloConnection
    let frameSource: TelloFrameSource
    private var obstacleGuard: ObstacleGuard?

    // MARK: - Configuration

    private let config: DroneConfig

    // MARK: - State streams

    /// Async stream of drone state changes, forwarded from TelloConnection.
    nonisolated let stateStream: AsyncStream<TelloState>

    /// Async stream of alerts for UI consumption.
    nonisolated let alertStream: AsyncStream<DroneAlert>
    private let alertContinuation: AsyncStream<DroneAlert>.Continuation

    // MARK: - Internal state

    private var monitorTask: Task<Void, Never>?
    private var searchTask: Task<Void, Never>?
    private var commandedLanding = false

    private(set) var searchActive = false
    private(set) var searchPaused = false
    private(set) var searchProgress: (current: Int, total: Int) = (0, 0)

    // MARK: - Init

    init(config: DroneConfig = DroneConfig()) {
        self.config = config
        self.wifiChecker = TelloWifiChecker(
            telloHost: config.telloDefaultHost,
            telloPort: config.telloCommandPort
        )
        self.connection = TelloConnection(config: config)
        self.frameSource = TelloFrameSource()

        // Forward connection's stateStream
        self.stateStream = connection.stateStream

        // Create alert stream
        var continuation: AsyncStream<DroneAlert>.Continuation!
        self.alertStream = AsyncStream { continuation = $0 }
        self.alertContinuation = continuation
    }

    deinit {
        alertContinuation.finish()
    }

    // MARK: - Connect / Disconnect

    /// Connect to the Tello, start video stream, begin telemetry monitoring,
    /// and initialize obstacle avoidance in the background.
    ///
    /// - Returns: `true` on successful connection and stream start.
    func connectAndStream() async -> Bool {
        // Step 1: Check WiFi
        let wifiStatus = await wifiChecker.check()
        guard case .onTelloWifi = wifiStatus else {
            let guidance = await wifiChecker.getGuidanceMessage(wifiStatus)
            Self.logger.warning("Not on Tello WiFi: \(guidance)")
            return false
        }

        // Step 2: Connect
        let connected = await connection.connect()
        guard connected else {
            Self.logger.error("Connection failed")
            return false
        }

        // Step 3: Start stream
        let streaming = await connection.startStream()
        guard streaming else {
            Self.logger.error("Stream start failed")
            await connection.disconnect()
            return false
        }

        // Step 4: Wire up frozen-stream recovery and start frame source
        frameSource.onStreamFrozen = { [weak self] in
            guard let self else { return }
            Task { await self.recoverStream() }
        }
        frameSource.start()

        // Step 5: Start telemetry monitoring
        startMonitoring()

        // Step 6: Initialize obstacle guard in background
        Task { [weak self] in
            guard let self else { return }
            await self.initObstacleGuard()
        }

        Self.logger.info("Connected and streaming")
        return true
    }

    /// Disconnect from the Tello. Lands the drone first if still flying
    /// (with a 3-second timeout).
    func disconnect() async {
        monitorTask?.cancel()
        monitorTask = nil

        // Land if flying
        let currentState = await connection.state
        if currentState.telemetry.isFlying {
            Self.logger.warning("Disconnecting while flying -- attempting to land first")
            do {
                try await withThrowingTaskGroup(of: Bool.self) { group in
                    group.addTask {
                        return await self.connection.land()
                    }
                    group.addTask {
                        try await Task.sleep(nanoseconds: 3_000_000_000)
                        throw CancellationError()
                    }
                    // Take whichever finishes first
                    _ = try await group.next()
                    group.cancelAll()
                }
            } catch {
                Self.logger.error("Land attempt during disconnect failed: \(error.localizedDescription)")
            }
        }

        // Tear down frame source
        frameSource.onStreamFrozen = nil
        frameSource.stop()

        // Stop stream (ignore errors)
        await connection.stopStream()

        // Disconnect connection
        await connection.disconnect()

        // Clean up obstacle guard
        obstacleGuard = nil

        Self.logger.info("Disconnected")
    }

    // MARK: - Obstacle Guard

    /// Initialize the obstacle avoidance system.
    private func initObstacleGuard() async {
        guard obstacleGuard == nil else { return }

        let guard_ = ObstacleGuard()
        await guard_.initialize()

        if await guard_.isAvailable {
            obstacleGuard = guard_
            Self.logger.info("Obstacle avoidance enabled")
        } else {
            Self.logger.warning("Obstacle avoidance: MiDaS model not available")
        }
    }

    // MARK: - Stream Recovery

    /// Recover a frozen video stream by cycling streamoff/streamon and
    /// restarting the frame source.
    private func recoverStream() async {
        Self.logger.warning("Recovering frozen stream")

        frameSource.stop()
        _ = await connection.sendCommand("streamoff")

        try? await Task.sleep(nanoseconds: 500_000_000) // 500ms

        _ = await connection.sendCommand("streamon")

        try? await Task.sleep(nanoseconds: 200_000_000) // 200ms

        frameSource.onStreamFrozen = { [weak self] in
            guard let self else { return }
            Task { await self.recoverStream() }
        }
        frameSource.start()

        alertContinuation.yield(.streamFrozen)
    }

    // MARK: - Lifecycle Hooks

    /// Called when the app enters the background. Hovers the drone if flying.
    func onScenePhaseBackground() async {
        let currentState = await connection.state
        if currentState.telemetry.isFlying {
            Self.logger.info("Scene phase background while flying -- hovering")
            await connection.hover()
        }
    }

    /// Called on app termination. Emergency lands and disconnects.
    func onTermination() async {
        monitorTask?.cancel()
        monitorTask = nil
        searchTask?.cancel()
        searchTask = nil

        frameSource.stop()

        let currentState = await connection.state
        if currentState.telemetry.isFlying {
            Self.logger.warning("Termination while flying -- emergency landing")
            // Attempt landing with short timeout
            _ = await withTaskGroup(of: Bool.self) { group in
                group.addTask {
                    return await self.connection.land()
                }
                group.addTask {
                    try? await Task.sleep(nanoseconds: 1_500_000_000)
                    return false
                }
                let result = await group.next() ?? false
                group.cancelAll()
                return result
            }
        }

        await connection.disconnect()
        obstacleGuard = nil
        Self.logger.info("Terminated")
    }

    // MARK: - Flight Command Delegates

    func takeoff() async -> Bool {
        await connection.takeoff()
    }

    func land() async -> Bool {
        commandedLanding = true
        return await connection.land()
    }

    func move(direction: String, distanceCm: Int) async {
        await connection.move(direction: direction, distanceCm: distanceCm)
    }

    func rotate(degrees: Int) async {
        await connection.rotate(degrees: degrees)
    }

    func rcControl(lr: Int, fb: Int, ud: Int, yaw: Int) async {
        await connection.rcControl(lr: lr, fb: fb, ud: ud, yaw: yaw)
    }

    func hover() async {
        await connection.hover()
    }

    nonisolated func emergencyStop() async -> Bool {
        // Set commandedLanding inside actor isolation first
        await setCommandedLanding(true)
        return await connection.emergencyStop()
    }

    /// Actor-isolated helper for emergencyStop.
    private func setCommandedLanding(_ value: Bool) {
        commandedLanding = value
    }

    // MARK: - Pause / Resume for Match Detection

    /// Pause the active search pattern and hover in place.
    /// Called when the vision pipeline detects a confident match.
    func pauseSearchForMatch() async {
        guard searchActive, !searchPaused else { return }
        searchPaused = true
        Self.logger.info("Search paused -- match detected, hovering")
        await connection.hover()
        alertContinuation.yield(.matchPause)
    }

    /// Resume the search pattern after the operator dismisses a match alert.
    func resumeSearch() {
        guard searchPaused else { return }
        searchPaused = false
        Self.logger.info("Search resumed")
    }

    // MARK: - Search Pattern Execution

    /// Start executing a search pattern with obstacle avoidance.
    ///
    /// Generates waypoints from `SearchPattern.generate(type:)` and executes
    /// them sequentially. Before each forward move, checks for obstacles and
    /// attempts up to 5 evasive maneuvers per waypoint.
    func startSearchPattern(_ pattern: PatternType = .expandingSquare) {
        guard !searchActive else { return }

        let waypoints = SearchPattern.generate(type: pattern)
        searchActive = true
        searchPaused = false
        searchProgress = (0, waypoints.count)
        Self.logger.info("Starting search pattern: \(pattern.rawValue), \(waypoints.count) waypoints")

        searchTask = Task { [weak self] in
            guard let self else { return }

            for (index, wp) in waypoints.enumerated() {
                let active = await self.isSearchActive()
                if !active { break }

                // Wait while paused for match inspection
                while true {
                    let paused = await self.isSearchPaused()
                    let active2 = await self.isSearchActive()
                    guard paused && active2 else { break }
                    try? await Task.sleep(nanoseconds: 200_000_000)
                }
                let stillActive = await self.isSearchActive()
                if !stillActive { break }

                await self.updateSearchProgress(current: index + 1, total: waypoints.count)

                do {
                    if wp.distanceCm > 0 {
                        // Obstacle check before each forward move
                        let pathClear = await self.checkAndAvoidObstacles(waypointIndex: index)
                        if !pathClear {
                            Self.logger.warning("Skipping waypoint \(index + 1) -- obstacle avoidance exhausted")
                            continue
                        }

                        await self.connection.move(direction: wp.direction, distanceCm: wp.distanceCm)
                        try await Task.sleep(nanoseconds: 500_000_000) // 500ms
                    }
                    if wp.rotateDegrees != 0 {
                        await self.connection.rotate(degrees: wp.rotateDegrees)
                        try await Task.sleep(nanoseconds: 500_000_000) // 500ms
                    }
                } catch is CancellationError {
                    break
                } catch {
                    Self.logger.error("Search pattern error at waypoint \(index + 1): \(error.localizedDescription)")
                    break
                }
            }

            await self.clearSearchState()
            Self.logger.info("Search pattern complete")
        }
    }

    /// Stop the active search pattern and hover.
    func stopSearchPattern() async {
        searchActive = false
        searchPaused = false
        searchTask?.cancel()
        searchTask = nil
        searchProgress = (0, 0)
        await connection.hover()
        Self.logger.info("Search pattern stopped")
    }

    // MARK: - Search State Helpers

    private func isSearchActive() -> Bool { searchActive }
    private func isSearchPaused() -> Bool { searchPaused }

    private func updateSearchProgress(current: Int, total: Int) {
        searchProgress = (current, total)
    }

    private func clearSearchState() {
        searchActive = false
        searchPaused = false
        searchProgress = (0, 0)
    }

    // MARK: - Obstacle Avoidance

    /// Check for obstacles before moving. If an obstacle is detected,
    /// perform evasive maneuvers up to `maxAvoidanceRetries` times.
    ///
    /// - Returns: `true` if the path is now clear, `false` if retries exhausted.
    private func checkAndAvoidObstacles(waypointIndex: Int) async -> Bool {
        guard let guard_ = obstacleGuard else { return true }

        for retry in 1...Self.maxAvoidanceRetries {
            guard searchActive else { return false }

            guard let frame = frameSource.getLatestFrame() else { return true }

            let check = await guard_.checkPath(frame: frame)
            if check.safe { return true }

            Self.logger.warning("Obstacle at waypoint \(waypointIndex + 1): action=\(check.action), depth=\(check.centerDepth), retry=\(retry)")
            alertContinuation.yield(.obstacleDetected(action: check.action, confidence: check.confidence))

            if retry >= Self.maxAvoidanceRetries {
                Self.logger.warning("Max avoidance retries at waypoint \(waypointIndex + 1)")
                return false
            }

            do {
                switch check.action {
                case "go_left":
                    await connection.move(direction: "left", distanceCm: Self.avoidanceSideCm)
                    try await Task.sleep(nanoseconds: 500_000_000)

                case "go_right":
                    await connection.move(direction: "right", distanceCm: Self.avoidanceSideCm)
                    try await Task.sleep(nanoseconds: 500_000_000)

                case "reverse":
                    await connection.move(direction: "back", distanceCm: Self.avoidanceReverseCm)
                    try await Task.sleep(nanoseconds: 500_000_000)
                    await connection.rotate(degrees: 90)
                    try await Task.sleep(nanoseconds: 500_000_000)

                default:
                    break
                }
            } catch is CancellationError {
                return false
            } catch {
                Self.logger.error("Avoidance maneuver failed: \(error.localizedDescription)")
                return false
            }
        }
        return false
    }

    // MARK: - Telemetry Monitoring

    /// Start monitoring telemetry for battery, crash, and connection-loss alerts.
    ///
    /// Mirrors the Android `startMonitoring()` logic:
    /// - Connection state changes trigger `.connectionLost`
    /// - Battery at/below `batteryWarnThreshold` triggers `.batteryWarning`
    /// - Battery at/below `batteryCriticalThreshold` triggers `.batteryCritical` + auto-land
    /// - isFlying going false while connected (not from commanded landing) triggers `.crashDetected`
    private func startMonitoring() {
        monitorTask?.cancel()
        monitorTask = Task { [weak self] in
            guard let self else { return }

            var wasConnected = true
            var wasFlying = false
            var batteryWarnSent = false
            var batteryCriticalSent = false

            for await state in await self.connection.stateStream {
                guard !Task.isCancelled else { break }

                // Connection loss detection
                let isConnected = state.connectionState != .disconnected &&
                    state.connectionState != .error
                if wasConnected && !isConnected {
                    Self.logger.error("Connection lost")
                    await self.emitAlert(.connectionLost)
                }
                wasConnected = isConnected

                // Battery monitoring
                if let battery = state.telemetry.battery {
                    let warnThreshold = await self.getWarnThreshold()
                    let criticalThreshold = await self.getCriticalThreshold()

                    if battery > warnThreshold {
                        batteryWarnSent = false
                        batteryCriticalSent = false
                    } else if battery > criticalThreshold {
                        batteryCriticalSent = false
                        if !batteryWarnSent {
                            Self.logger.warning("Battery warning: \(battery)%")
                            await self.emitAlert(.batteryWarning(battery))
                            batteryWarnSent = true
                        }
                    } else {
                        // battery <= criticalThreshold
                        if !batteryWarnSent {
                            Self.logger.warning("Battery warning: \(battery)%")
                            await self.emitAlert(.batteryWarning(battery))
                            batteryWarnSent = true
                        }
                        if !batteryCriticalSent {
                            Self.logger.warning("Battery critical: \(battery)%")
                            await self.emitAlert(.batteryCritical(battery))
                            batteryCriticalSent = true
                            if state.telemetry.isFlying {
                                Self.logger.warning("Auto-landing due to critical battery")
                                _ = await self.land()
                            }
                        }
                    }
                }

                // Crash detection
                if wasFlying && !state.telemetry.isFlying {
                    let didCommandLanding = await self.didCommandLanding()
                    if state.telemetry.height == 0 && isConnected && !didCommandLanding {
                        Self.logger.error("Crash detected")
                        await self.emitAlert(.crashDetected)
                    }
                    await self.resetCommandedLanding()
                }
                wasFlying = state.telemetry.isFlying
            }
        }
    }

    // MARK: - Monitoring Helpers

    private func emitAlert(_ alert: DroneAlert) {
        alertContinuation.yield(alert)
    }

    private func getWarnThreshold() -> Int { config.batteryWarnThreshold }
    private func getCriticalThreshold() -> Int { config.batteryCriticalThreshold }
    private func didCommandLanding() -> Bool { commandedLanding }
    private func resetCommandedLanding() { commandedLanding = false }
}
